File: Codechef/B/72/D.cpp
#pragma GCC optimize(1)
#pragma GCC optimize(2)
#pragma GCC optimize(3,"Ofast","inline")
#include<bits/stdc++.h>
using namespace std;

const long long mod = 1000000007;

long long quickpow(long long a, long long b, long long mod){
    long long ans;
    for (ans = 1; b; b >>= 1){
        if (b & 1){
            ans = (ans * a) % mod;
        }
        a = (a * a) % mod;
    }
    return ans;
}

template <class T>
void chmin(T &a, T b){
    if (b < a) a = b;
}
template <class T>
void chmax(T &a, T b){
    if (b > a) a = b;
}

const int inf = 0x3f3f3f3f;
const long long infl = 0x3f3f3f3f3f3f3f3f;

int dir[4][2] = {1, 0, -1, 0, 0, -1, 0, 1};

const int maxcnt = 15000010;
const int Size = 2;
struct Trie {
    int root, cnt;
    int tree[maxcnt][Size];
    int sz[maxcnt];
    int create() {
        ++cnt;
        memset(tree[cnt], 0, sizeof(tree[cnt]));
        sz[cnt] = 0;
        return cnt;
    }
    void Init() {
        cnt = 0;
        root = create();
    }

    void add(int num){
        int pos = root;
        sz[pos]++;
        for (int i = 30; i >= 0; i--){
            int w = num >> i & 1;
            if (tree[pos][w] == 0){
                tree[pos][w] = create();
            }
            pos = tree[pos][w];
            sz[pos]++;
        }
    }

    void del(int num){
        int pos = root;
        sz[pos]--;
        for (int i = 30; i >= 0; i--){
            int w = num >> i & 1;
            if (tree[pos][w] == 0){
                tree[pos][w] = create();
            }
            pos = tree[pos][w];
            sz[pos]--;
        }
    }
    long long query(int num){
        int pos = root;
        long long res = 0;
        for (int i = 30; i >= 0; i--){
            int w = num >> i & 1;
            if (tree[pos][1 - w] && sz[tree[pos][1 - w]]){
                res |= 1 << i;
                pos = tree[pos][1 - w];
            }
            else if (tree[pos][w] && sz[tree[pos][w]]){
                pos = tree[pos][w];
            }
            else{
                break;
            }
        }
        return res;
    }
};

Trie T;

const int N = 500010;
vector<int> adj[N], child[N];
long long prob[N], vals[N], A[N];
long long res;

void dfs(int node, int parent){
    for (auto& o: adj[node]){
        if (o == parent) continue;
        child[node].push_back(o);
        vals[o] = vals[node] ^ A[o];
        dfs(o, node);
    }
}

void dfs2(int node){
    if (node == 1){
        T.Init();
    }
    T.add(A[node]);
    int t = child[node].size();
    if (t){
        for (auto& o: child[node]){
            prob[o] = prob[node] * quickpow(t, mod - 2, mod) % mod;
            dfs2(o);
        }
    }
    else{
        res += prob[node] * T.query(vals[node]) % mod;
        res %= mod;
    }
    T.del(A[node]);
}

void solve(int testcase){
    int n;
    cin >> n;
    for (int i = 1; i <= n; i++){
        adj[i].clear();
        child[i].clear();
        cin >> A[i];
    }

    for (int i = 1; i < n; i++){
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    vals[1] = A[1];
    dfs(1, 0);

    prob[1] = 1LL;
    res = 0LL;
    dfs2(1);

    cout << res << "\n";
}

int main(){
    ios_base::sync_with_stdio(0); cin.tie(0); cout.tie(0);
    int T = 1;
    cin >> T;
    for (int testcase = 1; testcase <= T; testcase++) solve(testcase);
    return 0;
}

File: Codechef/B/72/D.py
'''
Hala Madrid!
https://www.zhihu.com/people/li-dong-hao-78-74
'''

import sys
import os
from io import BytesIO, IOBase
BUFSIZE = 8192
class FastIO(IOBase):
    newlines = 0
    def __init__(self, file):
        self._fd = file.fileno()
        self.buffer = BytesIO()
        self.writable = "x" in file.mode or "r" not in file.mode
        self.write = self.buffer.write if self.writable else None
    def read(self):
        while True:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            if not b:
                break
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines = 0
        return self.buffer.read()
    def readline(self):
        while self.newlines == 0:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            self.newlines = b.count(b"\n") + (not b)
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines -= 1
        return self.buffer.readline()
    def flush(self):
        if self.writable:
            os.write(self._fd, self.buffer.getvalue())
            self.buffer.truncate(0), self.buffer.seek(0)
class IOWrapper(IOBase):
    def __init__(self, file):
        self.buffer = FastIO(file)
        self.flush = self.buffer.flush
        self.writable = self.buffer.writable
        self.write = lambda s: self.buffer.write(s.encode("ascii"))
        self.read = lambda: self.buffer.read().decode("ascii")
        self.readline = lambda: self.buffer.readline().decode("ascii")
sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)
input = lambda: sys.stdin.readline().rstrip("\r\n")

def I():
    return input()
def II():
    return int(input())
def MI():
    return map(int, input().split())
def LI():
    return list(input().split())
def LII():
    return list(map(int, input().split()))
def GMI():
    return map(lambda x: int(x) - 1, input().split())

#------------------------------FastIO---------------------------------

from bisect import *
from heapq import *
from collections import *
from functools import *
from itertools import *
from time import *
from random import *
from math import log, gcd, sqrt, ceil

'''
手写栈防止recursion limit
注意要用yield 不要用return
函数结尾要写yield None
'''
from types import GeneratorType
def bootstrap(f, stack=[]):
    def wrappedfunc(*args, **kwargs):
        if stack:
            return f(*args, **kwargs)
        else:
            to = f(*args, **kwargs)
            while True:
                if type(to) is GeneratorType:
                    stack.append(to)
                    to = next(to)
                else:
                    stack.pop()
                    if not stack:
                        break
                    to = stack[-1].send(to)
            return to
    return wrappedfunc
    
class Trie:
    N = 30
    def __init__(self):
        self.left = None
        self.right = None
        self.cnt = 0
    
    def insert(self, x: int) -> None:
        root = self
        self.cnt += 1
        for i in range(Trie.N, -1, -1):
            ID = (x >> i) & 1
            if ID == 0:
                if root.left == None:
                    root.left = Trie()
                root = root.left
            else:
                if root.right == None:
                    root.right = Trie()
                root = root.right
            root.cnt += 1
    
    def delete(self, x: int) -> None:
        root = self
        self.cnt -= 1
        for i in range(Trie.N, -1, -1):
            ID = (x >> i) & 1
            if ID == 0:
                if root.left == None:
                    root.left = Trie()
                root = root.left
            else:
                if root.right == None:
                    root.right = Trie()
                root = root.right
            root.cnt -= 1
    
    def query_min(self, x: int) -> int:
        root = self
        res = 0
        for i in range(Trie.N, -1, -1):
            ID = (x >> i) & 1
            if ID == 0:
                if root.left != None:
                    root = root.left
                elif root.right != None:
                    res |= 1 << i
                    root = root.right
                else:
                    break
            else:
                if root.right != None:
                    root = root.right
                elif root.left != None:
                    res |= 1 << i
                    root = root.left
                else:
                    break
        return res
    
    def query_max(self, x: int) -> int:
        root = self
        res = 0
        for i in range(Trie.N, -1, -1):
            ID = (x >> i) & 1
            if ID == 0:
                if root.right != None and root.right.cnt:
                    res |= 1 << i
                    root = root.right
                elif root.left != None:
                    root = root.left
                else:
                    break
            else:
                if root.left != None and root.left.cnt:
                    res |= 1 << i
                    root = root.left
                elif root.right != None:
                    root = root.right
                else:
                    break
        return res

mod = 10 ** 9 + 7

@lru_cache(None)
def inv(num):
    return pow(num, mod - 2, mod)

def solve(testcase):
    n = II()
    A = LII()
    TRIE = Trie()
    adj = defaultdict(list)
    for _ in range(n - 1):
        u, v = MI()
        adj[u].append(v)
        adj[v].append(u)
    
    value = [-1 for _ in range(n + 1)]
    value[1] = A[0]
    child = defaultdict(list)
    @bootstrap
    def dfs(node, parent):
        for o in adj[node]:
            if o == parent:
                continue
            else:
                child[node].append(o)
                value[o] = value[node] ^ A[o - 1]
                yield dfs(o, node)
        yield None
    
    dfs(1, 0)

    prob = [-1 for _ in range(n + 1)]
    prob[1] = 1
    res = 0

    @bootstrap
    def dfs2(node):
        nonlocal res
        TRIE.insert(A[node - 1])
        if child[node]:
            p = prob[node] * inv(len(child[node])) % mod
            for o in child[node]:
                prob[o] = p
                yield dfs2(o)
        else:
            res += prob[node] * TRIE.query_max(value[node]) % mod
            res %= mod
        
        TRIE.delete(A[node - 1])
        yield None
    
    dfs2(1)

    print(res)


for testcase in range(II()):
    solve(testcase)